pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Producer.java

```diff
@@ -132,6 +132,21 @@ public interface Producer<T> extends Closeable {
      * @since 2.7.0
      */
     TypedMessageBuilder<T> newMessage(Transaction txn);
+
+    /**
+     * Create a new message builder with transaction and schema, not required same parameterized type with the
+     * producer.
+     *
+     * <p>After the transaction commit, it will be made visible to consumer.
+     *
+     * <p>After the transaction abort, it will never be visible to consumer.
+     *
+     * @return a typed message builder that can be used to construct the message to be sent through this producer
+     * @see #newMessage()
+     */
+    <V> TypedMessageBuilder<V> newMessage(Schema<V> schema,
+                                          Transaction txn);
```

Comment:
This change adds a new public API, which requires a PIP. https://github.com/apache/pulsar/tree/master/pip#when-is-a-pip-required

BTW, for this implementation PR, you should add a test to ensure it works.

Reply:
Is there a test suite for the producer? I was unable to find one to add to.

Reply:
No. Instead, I think you should add a test to `TransactionTest` to show which case this new API should be used. e.g.

```java
// producer is a Producer<String> while value is a String rather than byte[]
producer.newMessage(schema, txn).value(value).send();
```